drv/lpc55-spi-server/src/main.rs

```diff
@@ -8,24 +8,329 @@
+        // SP can issue a command (or no-op) at any start of transmission event.
+        if sot {
+            ringbuf_entry!(Trace::RxState(self.state)); // log old state
+            self.ready();
+        }
+
+        match self.state {
+            RxState::Invalid => {
+                // Ignoring bytes due to previous error or completion of Rx.
+            }
+            RxState::Idle => {
+                // The first received byte must be a supported protocol number.
+                self.rx[self.count] = b;
+                self.count += 1;
+                match b {
+                    SPI_MSG_IGNORE => {
+                        // A zero byte marks the message as one to ignore.
+                        self.state = RxState::Invalid;
+                    }
+                    SPI_MSG_VERSION => {
+                        // The only supported protocol at this time is 0x01.
+                        // Collect a header.
+                        self.state = RxState::Header;
+                    }
+                    _ => {
+                        // Form an error response unless Tx is sending something
+                        // valid.
+                        if tx.enqueue(
+                            MsgType::ErrorRsp,
+                            EnqueueBuf::Copy(&self.rx[0..1]),
+                        ) {
+                            ringbuf_entry!(Trace::Enqueued);
+                            self.state = RxState::Dispatch; // or RxState::Error?
+                                                            // TODO: a documented error type should be the first byte.
+                        } else {
+                            // We received garbage at sot but Tx was not idle.
+                            ringbuf_entry!(Trace::EnqueueFail);
+                        }
+                    }
+                }
+            }
+            RxState::Header => {
+                self.rx[self.count] = b;
+                self.count += 1;
+                ringbuf_entry!(Trace::HeaderCollect);
+                if self.count == SPI_HEADER_SIZE {
+                    ringbuf_entry!(Trace::HeaderComplete);
+                    let max_len = self.rx.len();
+                    let msg = Msg::parse(&mut *self.rx).unwrap_lite();
+                    self.end = msg.payload_len() + SPI_HEADER_SIZE;
+                    if self.end > max_len {
+                        ringbuf_entry!(Trace::BadMsgLength);
+                        self.state = RxState::Error;
+                        //let mut buf = [0u8; SPI_HEADER_SIZE];
+                        //buf[..SPI_HEADER_SIZE].clone_from_slice(&self.rx[..SPI_HEADER_SIZE]);
+
+                        if tx.enqueue(
+                            MsgType::ErrorRsp,
+                            EnqueueBuf::Copy(&self.rx[0..SPI_HEADER_SIZE]),
+                        ) {
+                            self.state = RxState::Dispatch;
+                        }
```

Comment:
What happens if we fail to enqueue an error message? In the equivalent case above, there's `ringbuf_entry!(Trace::EnqueueFail);`

Is it _correct_ to do nothing if we fail to enqueue an error message, vs waiting until Tx is idle then sending it? Will the host be waiting on a reply?